### PR TITLE
849 fix bug same primary keys on routepath creates invalid geom

### LINKED ISSUE #849: Create routePath allows to use links which geometries that are not linked to each other
BE: https://gitlab.hsl.fi/jore/jore-map-backend/merge_requests/57
FE: #856 

How to reproduce

Create new routePath with same start & end date as some other routePath X
--> new routePath has links from routePath X

* add validation + tests so that you can't create 2 routePaths with same id (check date)
* add validation so that startDate should be always before endDate

### PR DESCRIPTION
**Pull request by kronken**
Closes #849 

BE PR: https://gitlab.hsl.fi/jore/jore-map-backend/merge_requests/57

Test by trying to create routepath with same primary key as some other routepath

### CODE REVIEW

File: src/components/sidebar/routePathView/routePathInfoTab/RoutePathInfoTab.tsx [hunk 1]
```diff
@@ -1,54 +1,352 @@
...
+            this.props.setValidatorResult('direction', validationResult);
+        }
+    };
+
+    private validatePrimaryKey = (direction: string, startTime: Date) => {
+        this.props.onChange('direction')(direction);
+        this.props.onChange('startTime')(startTime);
+
+        if (this.props.isNewRoutePath) {
+            this.checkForDuplicatePrimaryKey();
```
Comment (Divinare) [inline src/components/sidebar/routePathView/routePathInfoTab/RoutePathInfoTab.tsx]: A bit cleaner code that I would use:
```
    private isPrimaryKeyDuplicated = () => {
        const routePath = this.props.routePathStore!.routePath!;

        return this.existingRoutePathPrimaryKeys.some(
            rp =>
                routePath.routeId === rp.routeId &&
                routePath.direction === rp.direction &&
                routePath.startTime.getTime() === rp.startTime.getTime()
        );
    };

    private validatePrimaryKey = (direction: string, startTime: Date) => {
        this.props.onChange('direction')(direction);
        this.props.onChange('startTime')(startTime);

        if (this.props.isNewRoutePath && this.isPrimaryKeyDuplicated()) {
            const validationResult: IValidationResult = {
                isValid: false,
                errorMessage:
                    'Reitinsuunta samalla reitillä, suunnalla ja alkupäivämäärällä on jo olemassa.'
            };
            this.props.setValidatorResult('direction', validationResult);
        }
    };
```

File: src/components/sidebar/routePathView/routePathInfoTab/RoutePathInfoTab.tsx [hunk 2]
```diff
@@ -1,54 +1,352 @@
...
+                            <InputContainer
+                                label='VOIM. AST'
+                                disabled={isUpdating}
+                                type='date'
+                                value={routePath.startTime}
+                                onChange={this.onChangeStartTime}
+                                validationResult={
+                                    invalidPropertiesMap['startTime']
+                                }
+                            />
```
Comment (Divinare) [inline src/components/sidebar/routePathView/routePathInfoTab/RoutePathInfoTab.tsx]: Hmm, maybe this inputContainer needs a new prop momentFormat?: string

then you could say:
momentFormat={'DD.MM.YYYY'}

and at inputContainer:
value={Moment(props.value).format(this.props.momentFormat ? momentFormat : 'DD.MM.YYYY HH:mm)}

(feel free to figure better variable name than momentFormat)

File: src/factories/routePathFactory.ts [hunk 3]
```diff
@@ -32,9 +43,14 @@ class RoutePathFactory {
         const exceptionPath = externalRoutePath.poikkeusreitti
             ? externalRoutePath.poikkeusreitti
             : '0';
+
+        const routePathPrimaryKey = RoutePathFactory.mapExternalRoutePathToRoutePathPrimaryKey(
+            externalRoutePath
+        );
+
         return {
+            ...routePathPrimaryKey,
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/factories/routePathFactory.ts]: sure, I can call them individually, maybe it is a bit clearer

File: src/components/sidebar/routePathView/routePathInfoTab/RoutePathInfoTab.tsx (hunk 1 above)
Comment (kronken) [reply src/components/sidebar/routePathView/routePathInfoTab/RoutePathInfoTab.tsx]: this is indeed nicer

**Comment by kronken** [comment]
Start and end date validation: https://github.com/HSLdevcom/jore-map-ui/issues/862

File: src/components/sidebar/routePathView/routePathInfoTab/RoutePathInfoTab.tsx (hunk 2 above)
Comment (kronken) [reply src/components/sidebar/routePathView/routePathInfoTab/RoutePathInfoTab.tsx]: https://github.com/HSLdevcom/jore-map-ui/issues/863

**Comment by kronken** [comment]
PR comments are now fixed, and this could be reviewed again.

**Comment by Divinare** [comment]
Review ok.

Only problem is, saving of routePath loses all routePath links after save & refreshing page

**Comment by Divinare** [comment]
Refreshing page when creating new routePath causes some console error

**Comment by kronken** [comment]
The warning message is now fixed

**Pull request merged**